### PR TITLE
added login and signup

### PR DESCRIPTION
**Pull request by incogni23**

### CODE REVIEW

File: auth/dao.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+// connect w db
+// credentials
+
+// query: update table
+// existing user check - where usrn = " given",
+
+//if fail : add a new record (user))- insert
+
+//else : err
```
Comment (shubham-aurva) [inline auth/dao.go]: remove them

File: auth/dao.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+	database gorm.DB
+}
+
+func NewDatabase(db gorm.DB) Dao {
+	return &dao{
+		database: db,
+	}
+}
+
+func (db *dao) InsertUser(u User) error {
```
Comment (shubham-aurva) [inline auth/dao.go]: variable name shouldn't be just `u`

File: auth/dao.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	return nil
+}
+
+func (db *dao) GetUser(username string) (*User, error) {
+	var user User
+	getUser := db.database.First(&user, "username = ?", username)
+	if getUser.Error != nil {
+		return nil, getUser.Error
+	}
+	return &user, nil
```
Comment (shubham-aurva) [inline auth/dao.go:46]: try to put some extra lines before a block for better readability

File: auth/service.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	"time"
+
+	"github.com/create"
+	"github.com/validate"
+)
+
+type User struct {
+	Username string
+	Password string
+}
```
Comment (shubham-aurva) [inline auth/service.go]: put it into a different file `models.go`

File: auth/service.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+type AuthServicer interface {
+	SignUp(u User) error
+	Login(username, password string) (string, error)
+	LoginWithToken(tokenString string) (string, error)
```
Comment (shubham-aurva) [inline auth/service.go:34]: why do we need this?

File: auth/service.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+func NewDBVar(d Dao) AuthServicer {
+	return &DBVar{
+		db: d,
+	}
+}
+
+// new signup
+
+func (dbv *DBVar) SignUp(u User) error {
```
Comment (shubham-aurva) [inline auth/service.go]: variable name should be proper

File: auth/service_test.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+		dbmock.On("InsertUser", mockUser).Return(nil)
+		dbmock.On("GetUser", "Ankita").Return(&blankUser, nil)
+
+		authService := auth.NewDBVar(dbmock)
+		err := authService.SignUp(mockUser)
+
+		assert.NoError(t, err)
+
+	})
```
Comment (shubham-aurva) [inline auth/service_test.go]: Wrong test cases. Should test login functionality, not the signup

File: auth/service.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	return "valid token", nil
+}
+
+// signup:
+//u1 User:
+// u1-> db
+// succes -> if user is suceesfully created in db as well
+// db X - fail
+
+//Login
```
Comment (shubham-aurva) [inline auth/service.go]: remove

**Review by shubham-aurva — CHANGES_REQUESTED** [review]

**Review by mishtuu — APPROVED** [review]

**Pull request merged**